src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +112,226 @@ public class FromBase64Transform : ICryptoTransform
     {
         private byte[] _inputBuffer = new byte[4];
         private int _inputIndex;
-        private FromBase64TransformMode _whitespaces;
+        private readonly FromBase64TransformMode _whitespaces;
 
         public FromBase64Transform() : this(FromBase64TransformMode.IgnoreWhiteSpaces) { }
         public FromBase64Transform(FromBase64TransformMode whitespaces)
         {
             _whitespaces = whitespaces;
-            _inputIndex = 0;
         }
 
         // Converting from Base64 generates 3 bytes output from each 4 bytes input block
+        private const int Base64InputBlockSize = 4;
         public int InputBlockSize => 1;
         public int OutputBlockSize => 3;
         public bool CanTransformMultipleBlocks => false;
         public virtual bool CanReuseTransform => true;
 
         public int TransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount != InputBlockSize is allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+
+            if (_inputBuffer == null)
+                ThrowHelper.ThrowObjectDisposed();
 
-            int effectiveCount;
-            byte[] temp = GetTempBuffer(inputBuffer, inputOffset, inputCount, out effectiveCount);
+            if (outputBuffer == null)
+                ThrowHelper.ThrowArgumentNull(ThrowHelper.ExceptionArgument.outputBuffer);
 
-            if (effectiveCount + _inputIndex < 4)
+            // The common case is inputCount = InputBlockSize
+            byte[] tmpBufferArray = null;
+            Span<byte> tmpBuffer = stackalloc byte[InputBlockSize];
+            if (inputCount > InputBlockSize)
```

Comment:
InputBlockSize (the property) is 1.  While that's the value that CryptoStream is going to use, I wonder if this wouldn't be better as

```C#
Span<byte> tmpBuffer = stackalloc byte[64];

if (inputCount > tmpBuffer.Length)
{
    tmpBuffer = tmpBufferArray = CryptoPool.Rent(inputCount);
}

...
```

The constant could be anything, really.  The question is more: Does it hurt if this allows more than one "InputBlockSize" in the stack "rent"? If no, does it hurt if it allows more than one "Base64InputBlockSize"? If no, let's avoid renting for longer.

Reply:
Ah, good point. This is more or less a leftover from my `InputBlockSize => 4` attempt, as than it would be a good match for the use with `CryptoStream`.

I'll do some experiments if the stackalloc can get bigger (64, 128) without perf-regression.

Reply:
According to my tests a value of `32` looks good.
([results](https://github.com/gfoidl/Benchmarks/blob/a3db43ef9067e54ed60a5e838e603ea7a7a196a6/corefx/System/Security/Cryptography/Base64TransformsBenchmarks/results/FromBase64TransformBlockBenchmark-report.md), the suffixes like `_4` indicate the buffer size that is stackalloced).

With `32` the common use cases are covered with stack allocation, and the JIT performs a optimization where this [stack allocation is treated as a local](https://github.com/dotnet/coreclr/pull/14623), so codegen is really good and is more or less just a stack-pointer move.
In this case theory is matched by benchmark-results 😃 

Larger values like 64, 128 for stack allocation don't show any benefit for this use case.
(and they result in a tight zeroing loop on the stack (which will be really performant if the size is small enough so that the loop stream detector of modern cpus can take advantage of this)).